### DISCUSSION TITLE
GW-BASIC is opensource...?

### DISCUSSION THREAD

**Discussion by cdaters** [picocomputer/community#110] [Q&A]
I just stumbled across this project yesterday and binge-watched the YouTube videos on Rumbledethumps' YouTube channel. I am very intrigued by the notion of this project and am considering building one myself. I must admit that I am not an Assembly person, but more of a script-kiddie with higher-level languages and I got my start with the venerable VIC-20 and C64 BASIC in junior high, before hitting the ground running with GW-BASIC on a TRS-80 IV computer back in high school. In the video on BASIC, I watched as Rumbledethumps discussed and showcased EhBASIC and touched on the reasons for its inclusion.

That said, I see that [GW-BASIC was open-sourced in 2020](https://devblogs.microsoft.com/commandline/microsoft-open-sources-gw-basic/) and is available on [GitHub here](https://github.com/microsoft/GW-BASIC). I'd never heard of EhBASIC before this project and I am wondering if GW-BASIC would be better? There is even another Microsoft Basic for the 6502 available on [GitHub here](https://github.com/mist64/msbasic).

**Comment by ceptimus** [comment]
As far as I know, the open source for GW-BASIC is in 8088 (PC) Assembler format, and there isn't an open source version for the 6502.

EhBASIC seems to be based on a modified and commented disassembly of Microsoft Basic: there might even be copyright problems with that in the unlikely event that someone writes a highly profitable application based on it!

BBC Basic is very powerful, and was originally written for the 6502 (mainly by Sophie (previously Roger) Wilson at Acorn for the BBC Microcomputer).  Whether or not the 6502 versions of it are open source now is also (as far as I know) open to debate, though the converted version for Z80 apparently is.  There are lots of 6502 versions on GitHub for other 6502 computers, so I think it's extremely unlikely there would be any copyright claims if someone ported BBC BASIC to the picocomputer.  It would likely perform much better than EhBASIC, if it were available, but it's a lot of work for someone to do.

**Reply by cdaters** [reply]
Very informative...thank you! Yeah, I came into all of this at the 8088 level, so all this 6502 stuff is very interesting to me. It even has me revisiting some of my BASIC stuff from the 80s, lol.

**Comment by rumbledethumps** [comment]
Is there any evidence for EhBASIC being based on a disassembly of MS BASIC? Or a citation?

**Reply by ceptimus** [reply]
There's a thread about it over at 6502.org [http://forum.6502.org/viewtopic.php?f=5&t=6194](http://forum.6502.org/viewtopic.php?f=5&t=6194)
John West's post, towards the end of that thread, gives a comparison disassembly of a small part of EhBASIC and Microsoft's equivalent.
He argues (pretty convincingly, in my opinion) that they are too similar for it just to be a coincidence, and speculates that the actual disassembly it came from might be the Apple II version.

There are differences though - I suppose it comes down to where you draw the line between 'inspired by' and 'copied from'.

**Reply by rumbledethumps** [reply]
The "LAB" disassembly looks like a recovery of lost source code and the supposedly similar opcodes don't look copied. It's certainly very possible Lee was referencing another BASIC while writing his, but the evidence presented is extremely weak. I'd want a lot more evidence before I repeat someone's claim that code was copied without attribution or permission.

**Comment by ceptimus** [comment]
I think Lee put 'LAB_' as the first four characters of many of the labels in his source.  You can see it in the archived copy of his own website [where he's documenting useful routines](http://www.6502.org/users/mycorner/6502/ehbasic/routines.html) that you might wish to call from your own assembly routines.

**Reply by rumbledethumps** [reply]
The LAB_ comes from a disassembler. My guess is that Lee lost the original source code or it was so unsuitable for a modern assembler that a disassembly was easier or safer than migrating.
